File: v2/144.py



# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def preorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        # 4 star, 必须熟练
        rs = []
        self.preorder(root, rs)
        return rs
    def preorder(self, node, rs):
        if node:
            rs.append(node.val)
            self.preorder(node.left, rs)
            self.preorder(node.right, rs)




class Solution(object):
    def preorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        # 6 star, 递归版很简单，迭代版没写出来
        # 使用栈，先入右子树，再入左子树，迭代时后入先出，因此会先迭代左子树
        if not root:
            return []
        stack = [root]
        ret = []
        while stack:
            cur = stack.pop()
            ret.append(cur.val)
            if cur.right:
                stack.append(cur.right)
            if cur.left:
                stack.append(cur.left)
        return ret

File: go/144.go
//package main
//
//import "fmt"

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

//type TreeNode struct {
//	Val   int
//	Left  *TreeNode
//	Right *TreeNode
//}

//5 star, 前序遍历，先本身，再左，再右
func preorderTraversal(root *TreeNode) []int {
	if root == nil {
		return []int{}
	}
	if root.Right == nil && root.Left == nil {
		return []int{root.Val}
	}
	var nums []int
	return preorder(root, nums)

}

func preorder(node *TreeNode, nums []int) []int {
	nums = append(nums, node.Val)

	if node.Left != nil {
		nums = preorder(node.Left, nums)
	}
	if node.Right != nil {
		nums = preorder(node.Right, nums)
	}
	return nums

}


// 非递归实现前序遍历，使用栈，先将根节点入栈，随后从栈中取出节点，如果该节点有右子树，将右子树入栈，如果该节点有左子树，将左子树入栈
func preorderTraversal2(root *TreeNode) []int {
	if root == nil {
		return []int{}
	}
	stack := []*TreeNode{root}
	var result []int
	var node *TreeNode
	for len(stack) > 0 {
		node, stack = stack[len(stack)-1], stack[:len(stack)-1]
		result = append(result, node.Val)
		if node.Right != nil {
			stack = append(stack, node.Right)
		}
		if node.Left != nil {
			stack = append(stack, node.Left)
		}

	}
	return result

}




//func main() {
//	a := TreeNode{1, nil, nil}
//	b := TreeNode{2, nil, nil}
//	c := TreeNode{3, nil, nil}
//	a.Right = &b
//	a.Left = &c
//	fmt.Println(preorderTraversal(&a))
//}
